File: bindings/rust/src/context.rs
use crate::ffi::{createContext, removeContext, contextError};

/// Represents a context for SOP operations.
///
/// The context is used to manage the lifecycle of operations and handle errors.
pub struct Context {
    /// The context ID.
    pub id: i64,
}

impl Context {
    /// Creates a new context.
    pub fn new() -> Self {
        unsafe {
            let id = createContext();
            Context { id }
        }
    }

    /// Checks for errors in the context.
    ///
    /// # Returns
    ///
    /// An option containing the error message if an error occurred, or `None` otherwise.
    pub fn error(&self) -> Option<String> {
        unsafe {
            let ptr = contextError(self.id);
            let res = crate::utils::process_go_result(ptr);
            if let Some(err_str) = res {
                if err_str.is_empty() {
                    None
                } else {
                    Some(err_str)
                }
            } else {
                None
            }
        }
    }
}

impl Drop for Context {
    fn drop(&mut self) {
        unsafe { removeContext(self.id) };
    }
}

impl Default for Context {
    fn default() -> Self {
        Self::new()
    }
}

File: bindings/python/sop/context.py
from . import call_go


class Context:
    """
    Represents a context for SOP operations.

    The context is used to manage the lifecycle of operations, handle errors, and allow for cancellation.
    It serves as a bridge between the Python client and the underlying Go implementation.
    """

    def __init__(self):
        """
        Initializes a new Context.

        This creates a corresponding context in the Go runtime and stores its ID.
        """
        self._removed = False
        self.id = call_go.create_context()

    def __del__(self):
        """
        Clean up the context.

        Removes the corresponding context from the Go runtime.
        """
        if not self._removed:
            self._removed = True
            call_go.remove_context(self.id)

    def cancel(self):
        """
        Cancels any running operation associated with this context.

        This signals the underlying Go runtime to abort the operation.
        """
        call_go.cancel_context(self.id)
        self._removed = True

    def error(self) -> str:
        """
        Checks for errors in the context.

        Returns:
            str: The error message if an error occurred (e.g., timeout or canceled), otherwise None.
        """
        return call_go.context_error(self.id)

    def is_valid(self) -> bool:
        """
        Checks if the context is valid.

        Returns:
            bool: True if the context has no errors, False otherwise.
        """
        return self.error() is None
